Add new functionality to kv_blockchain_db_editor

yontyon opened the pull request:
We add the following functionality to kv_blockchain_db_editor:
1. getEarliestCategoryUpdates - to get the earliest updates of a given category between genesis and an (optionally) given block
2. getCategoryEarliestStale - to get the earliest stale keys (grouped in a block) of a given category between  genesis and an (optionally) given block 
These two new functionalities will help us to test pruning

KTeofanidis commented on kvbc/tools/db_editor/include/kv_blockchain_db_editor.hpp:
```diff
@@ -301,6 +301,43 @@ struct GetCategories {
+  const bool read_only = true;
+  std::string description() const {
+    return "getLatestCategoryInfo CATEGORY-ID [BLOCK-VERSION]\n"
+           "  Returns the latest blockID that contains the givenb category, started from BLOCK-VERSION. If "
```
givenb typo.

yontyon replied:
fixed

KTeofanidis commented on kvbc/tools/db_editor/include/kv_blockchain_db_editor.hpp:
```diff
@@ -301,6 +301,43 @@ struct GetCategories {
+  std::string description() const {
+    return "getLatestCategoryInfo CATEGORY-ID [BLOCK-VERSION]\n"
+           "  Returns the latest blockID that contains the givenb category, started from BLOCK-VERSION. If "
+           "BLOCK-VERSION is not set, we start from the lastReachableBlock";
```
Maybe we should have a "." at the end

yontyon replied:
done

KTeofanidis commented on kvbc/tools/db_editor/include/kv_blockchain_db_editor.hpp:
```diff
@@ -301,6 +301,43 @@ struct GetCategories {
+      throw NotFoundException{"No Category ID was given"};
+    }
+    auto latestBlockID = adapter.getLastReachableBlockId();
+    if (args.values.size() == 2) {
```
In all other cases we have >= for the number of arguments. Maybe it should be the same here?

yontyon replied:
done

KTeofanidis commented:
We'll also need to add a test in kv_blockchain_db_editor_test.cpp

dartdart26 commented on kvbc/tools/db_editor/include/kv_blockchain_db_editor.hpp:
```diff
@@ -301,6 +301,43 @@ struct GetCategories {
   }
 };
 
+struct GetLatestCategoryInfo {
```
I was thinking a clearer name may be `GetLatestBlockIDContainingCategory`. What do you think?

dartdart26 replied:
Ah, we also return the updates themselves. So maybe then `GetLatestCategoryUpdates`?

yontyon replied:
done

dartdart26 commented on kvbc/tools/db_editor/include/kv_blockchain_db_editor.hpp:
```diff
@@ -301,6 +301,43 @@ struct GetCategories {
+struct GetLatestCategoryInfo {
+  const bool read_only = true;
+  std::string description() const {
+    return "getLatestCategoryInfo CATEGORY-ID [BLOCK-VERSION]\n"
```
Nit: Maybe we can rename to `BLOCK-VERSION-FROM`, i.e. `...contains the given category, starting from BLOCK-VERSION-FROM`.

dartdart26 replied:
Based on a comment below, maybe that should be called `BLOCK-VERSION-TO`. Did I get it correctly?

yontyon replied:
done

dartdart26 commented on kvbc/tools/db_editor/include/kv_blockchain_db_editor.hpp:
```diff
@@ -301,6 +301,43 @@ struct GetCategories {
+    auto cat = args.values.front();
+    std::map<std::string, std::string> cat_updates_map;
+    BlockId relevantBlockId = adapter.getGenesisBlockId();
+    for (auto block = adapter.getGenesisBlockId(); block <= latestBlockID; block++) {
```
Actually, I might have gotten the command wrong. Do we want to always search from genesis to some block ID? Or we want to search from some block ID to the end?

yontyon replied:
done

dartdart26 commented:
Can we also add a unit test for this new command?

yontyon commented:
@KTeofanidis , @dartdart26, notice that I've added one more functionality (to get the earliest stale keys of a category).
I've also added some basic tests

dartdart26 commented on kvbc/src/categorization/kv_blockchain.cpp:
```diff
@@ -320,6 +320,26 @@ std::optional<Updates> KeyValueBlockchain::getBlockUpdates(BlockId block_id) con
+  }
+
+  std::unordered_map<std::string, std::vector<std::string>> stale_keys;
+  // Iterate over groups and call corresponding deleteGenesisBlock,
+  // Each group is responsible to put its deletes into the batch
```
I think this comment is not relevant here.

yontyon replied:
fixed

dartdart26 commented on kvbc/test/sparse_merkle_storage/kv_blockchain_db_editor_test.cpp:
```diff
@@ -435,6 +435,64 @@ TEST_F(DbEditorTests, get_block_key_values) {
   ASSERT_THAT(out_.str(), EndsWith("\n}\n"));
 }
 
+TEST_F(DbEditorTests, get_latest_category_updates_merkle) {
+  ASSERT_EQ(EXIT_SUCCESS,
+            run(
+                CommandLineArguments{
+                    {kTestName, rocksDbPath(main_path_db_id_), "getEarliestCategoryUpdates", kCategoryMerkle}},
```
I think the test name is incorrect, because the command we call is `getEarliestCategoryUpdates()` and the test is called `latest`.

yontyon replied:
fixed

dartdart26 commented on kvbc/src/categorization/versioned_kv_category.cpp:
```diff
@@ -170,6 +170,21 @@ std::unordered_map<BlockId, std::vector<std::string>> VersionedKeyValueCategory:
+  for (const auto &[key, flags] : out.keys) {
+    const auto latest = getLatestVersion(key);
+    ConcordAssert(latest.has_value());
+    ConcordAssertEQ(latest->deleted, flags.deleted);
```
Based on the merkle category code, the assert we use there also makes sense here:
```
ConcordAssertLE(block_id, tagged_version->version); -> ConcordAssertLE(block_id, latest->version);
```

yontyon replied:
fixed

dartdart26 commented on kvbc/src/categorization/block_merkle_category.cpp:
```diff
@@ -541,6 +541,35 @@ std::pair<SetOfKeyValuePairs, KeysVector> BlockMerkleCategory::rewriteAlreadyPru
+
+    if (block_id == tagged_version->version) {
+      if (tagged_version->deleted) {
+        // The latest version is a tombstone. We can delete the key and version.
```
I think the `we can delete the key and version` part of the comment is not relevant here.

yontyon replied:
fixed

dartdart26 commented on kvbc/include/categorization/immutable_kv_category.h:
```diff
@@ -48,6 +48,7 @@ class ImmutableKeyValueCategory {
   // Adding keys that already exist in this category is undefined behavior.
   ImmutableOutput add(BlockId, ImmutableInput &&, storage::rocksdb::NativeWriteBatch &);
 
+  std::vector<std::string> getBlockStaleKeys(BlockId, const ImmutableOutput &);
```
Nit: Can put an empty line after the declaration.

dartdart26 replied:
Also, can make all these methods in all cats `const`.

yontyon replied:
Fixed the empty line.
As for the const issue, it requires to make other method const and I would prefer to avoid changing this

dartdart26 commented on kvbc/include/categorization/kv_blockchain.h:
```diff
@@ -64,6 +64,18 @@ class KeyValueBlockchain {
   std::optional<Hash> parentDigest(BlockId block_id) const;
   bool hasBlock(BlockId block_id) const;
 
+  std::vector<std::string> getStaleKeys(BlockId block_id,
+                                        const std::string& category_id,
+                                        const ImmutableOutput& updates_info);
+
+  std::vector<std::string> getStaleKeys(BlockId block_id,
+                                        const std::string& category_id,
+                                        const VersionedOutput& updates_info);
+
+  std::vector<std::string> getStaleKeys(BlockId block_id,
+                                        const std::string& category_id,
+                                        const BlockMerkleOutput& updates_info);
+
```
AFAIK, these can all be made `const`.

yontyon replied:
see my comment above

dartdart26 commented on kvbc/include/categorization/kv_blockchain.h:
```diff
@@ -90,6 +102,7 @@ class KeyValueBlockchain {
   // Get the updates that were used to create `block_id`.
   std::optional<Updates> getBlockUpdates(BlockId block_id) const;
 
+  std::unordered_map<std::string, std::vector<std::string>> getBlockStaleKeys(BlockId block_id);
```
Can be made const.

Also, maybe we can put a comment to explain what the output map contains (i.e. category ID -> stale keys).

Finally, maybe we can use `std::map` for uniformity and deterministic outputs between invocations. I think it makes sense to do that.

dartdart26 commented on kvbc/include/categorization/versioned_kv_category.h:
```diff
@@ -46,6 +46,7 @@ class VersionedKeyValueCategory {
 
   VersionedOutput add(BlockId, VersionedInput &&, storage::rocksdb::NativeWriteBatch &);
 
+  std::vector<std::string> getBlockStaleKeys(BlockId, const VersionedOutput &);
```
Can be made const.

dartdart26 commented on kvbc/tools/db_editor/include/kv_blockchain_db_editor.hpp:
```diff
@@ -301,6 +301,93 @@ struct GetCategories {
+struct GetEarliestCategoryUpdates {
+  const bool read_only = true;
+  std::string description() const {
+    return "getEarliestCategoryUpdates CATEGORY-ID [BLOCK-VERSION-TO]\n"
+           "  Returns the latest blockID that contains the given category, started from BLOCK-VERSION-TO. If "
+           "BLOCK-VERSION-TO is not set, we start from the lastReachableBlock.";
```
I think the description is confusing, because we say `earliest` and then `latest blockID`. Also, we say `started from`, but the block is `TO`.

Please correct me if I am wrong, but I think it can be described similar to:

```
Returns the first blockID that contains the given category in the [genesisBlockID, BLOCK-VERSION-TO] range.
If BLOCK-VERSION-TO is not set, the search range is [genesisBlockID, lastReachableBlockID].
```

dartdart26 commented on kvbc/tools/db_editor/include/kv_blockchain_db_editor.hpp:
```diff
@@ -301,6 +301,93 @@ struct GetCategories {
+struct GetCategoryEarliestStale {
+  const bool read_only = true;
+  std::string description() const {
+    return "getCategoryEarliestStale CATEGORY-ID [BLOCK-VERSION-TO]\n"
+           "  Returns the latest blockID that contains the given category, started from BLOCK-VERSION-TO. If "
+           "BLOCK-VERSION-TO is not set, we start from the lastReachableBlock.";
```
I think the description is incorrect.

Moreover, I guess we can call this command `GetCategoryStaleKeys` or similar. Also, the block parameter can be called just `BLOCK-ID` or `AS-OF-BLOCK-ID` as we want these stale keys per block.

dartdart26 replied:
I think I've misunderstood the idea behind that command. It is basically looking for the first block the category has at least one stale key? Could you please explain its use case?

yontyon replied:
discussed offline

dartdart26 commented on kvbc/tools/db_editor/include/kv_blockchain_db_editor.hpp:
```diff
@@ -301,6 +301,93 @@ struct GetCategories {
+  }
+};
+
+inline std::string getStaleKeysStr(std::vector<std::string> stale_keys) {
```
Nit: I think the parameter can be taken by const ref.

dartdart26 commented on kvbc/tools/db_editor/include/kv_blockchain_db_editor.hpp:
```diff
@@ -301,6 +301,93 @@ struct GetCategories {
+      stale_keys = adapter.getBlockStaleKeys(block);
+      if (stale_keys.count(cat) && !stale_keys[cat].empty()) {
+        relevantBlockId = block;
+        keys_as_string = getStaleKeysStr(stale_keys[cat]);
```
I was thinking if this can be represented as a JSON array instead. What do you think?

yontyon replied:
the "getStaleKeysStr" returns "[key1, key2,...keyn]"

dartdart26 commented on kvbc/test/sparse_merkle_storage/kv_blockchain_db_editor_test.cpp:
```diff
@@ -435,6 +435,64 @@ TEST_F(DbEditorTests, get_block_key_values) {
+  ASSERT_TRUE(err_.str().empty());
+  ASSERT_THAT(out_.str(), StartsWith("{\n  \"blockID\":"));
+  ASSERT_THAT(out_.str(), HasSubstr("\"category\": \"" + kCategoryImmutable + "\""));
+  ASSERT_THAT(out_.str(), HasSubstr("\"0x696d6d757461626c655f6b657931\": \"0x696d6d757461626c655f76616c31\""));
```
Is that the `immutable_key1` key and its value? If yes, maybe we can put a comment for clarity.

dartdart26 reviewed (approved):
Thanks @yontyon, it looks good!

I have added a few minor comments, mostly related to naming.

One thing that we may consider changing is to use a JSON array for the list of stale keys instead of a comma-separated list.
